### PR TITLE
CI: Implementing benchmark regression tests

### LINKED ISSUE #166: CI: Continuous Benchmarking
[github-action-benchmark](https://github.com/benchmark-action/github-action-benchmark) provides a [GitHub Action](https://github.com/features/actions) for continuous benchmarking, and we can utilize it to track performance regressions.

Expected output:
* Integrate [github-action-benchmark](https://github.com/benchmark-action/github-action-benchmark) in CI pipeline.
* Deploy [self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners) for stable benchmarking environments. Arm64 and x86-64 hosts included.
* Ensure that all benchmark charts from above workflows are gathered in GitHub pages.

Reference:
* [How to make GitHub Actions 22x faster with bare-metal Arm](https://actuated.dev/blog/native-arm64-for-github-actions)

### PR DESCRIPTION
**Pull request by ChinYikMing**
Using "pull_request_target" is required since benchmark action updates data in the "gh-pages" branch for visualization in gitHub pages, and "pull_request" does not grant the "GITHUB_TOKEN" needed for that purpose.

We also want "workflow_dispatch" because we require a method for running the benchmark the first time so that we can store the base benchmark and utilize it for benchmark comparison in the future.

Before merging this PR, please do the following:
1. Create a branch called "gh-pages", this is the default branch for gitHub repository to deploy our web-based  benchmark visualization. This branch can be left only ".git" directory.
2. Go to the setting of  repository  > Action > General > make sure that "read and write" permission is granted in "Workflow permissions" section

After merging this PR, please run the workflow called `Benchmark` manually via "Run workflow" button to generate the base benchmark. Note that it should be used in a reasonable timing in order not to mess up the benchmark visualization afterwards.

Last, the benchmark CI should work!

Close #166

### CODE REVIEW

File: .github/workflows/main.yml
Comment (jserv) [inline .github/workflows/main.yml]: Move benchmark part out of `main.yml` since the latter is complex enough.

**Comment by ChinYikMing** [comment]
One more thing: you can setup the self-hosted runner for consistency of performance(hopefully) afterwards since I do not have permission to do that.

**Comment by ChinYikMing** [comment]
[e829d89](https://github.com/sysprog21/rv32emu/pull/213/commits/e829d89661ff702ad976cf161c3922756876b8ce) changes the following:

1. Since the benchmark action has already been executed once when a pull request is opened, we also need to filter out the merging push event.
2. Run the "changed-files" action in higher order

For further merging the pull requests, please do not modify the commit message "Merge pull request " since it is the conditional of benchmark action not to run again after merging.

**Comment by jserv** [comment]
> Using "pull_request_target" is required since benchmark action updates data in the "gh-pages" branch for visualization in gitHub pages, and "pull_request" does not grant the "GITHUB_TOKEN" needed for that purpose.

Can we refrain from creating a `gh-pages` branch in this repository? I think this repository should remain as minimal as possible, primarily preserving git objects for source code and necessary materials. If needed, we can have a `gh-pages` branch in another repository.

**Comment by ChinYikMing** [comment]
> > Using "pull_request_target" is required since benchmark action updates data in the "gh-pages" branch for visualization in gitHub pages, and "pull_request" does not grant the "GITHUB_TOKEN" needed for that purpose.
> 
> Can we refrain from creating a `gh-pages` branch in this repository? I think this repository should remain as minimal as possible, primarily preserving git objects for source code and necessary materials. If needed, we can have a `gh-pages` branch in another repository.

Sure, a branch can be called whatever in another repository, I think the branch can be called as `rv32emu_benchmark` which makes more sense. 

Please give me URL of the repository that used to store the benchmark data.

**Comment by jserv** [comment]
> Please give me URL of the repository that used to store the benchmark data.

Let's try this: https://github.com/sysprog21/rv32emu-bench

**Comment by ChinYikMing** [comment]
Please add a secret called `RV32EMU_BENCH_TOKEN`(other naming is OK, if changes to other please inform me) to `rv32emu`  repository, this token should be a PAT that having permission in scope of repo.

**Comment by ChinYikMing** [comment]
Before merging this PR, go to the setting of `rv32emu` repository > Action > General > make sure that "read and write" permission is granted in "Workflow permissions" section

After merging this PR, please run the workflow called `Benchmark` manually via "Run workflow" button to generate the base benchmark. Note that it should be used in a reasonable timing in order not to mess up the benchmark visualization afterwards.

Now, the benchmark CI should work and benchmark data are stored in `rv32emu-bench` repository.

[Link of benchmark data visualization](https://sysprog21.github.io/rv32emu-bench/)

File: .github/workflows/benchmark.yml
Comment (jserv) [inline .github/workflows/benchmark.yml]: Replace `v3.6.0` with recent `v4`.

File: .github/workflows/benchmark.yml
Comment (jserv) [inline .github/workflows/benchmark.yml]: It is not necessary to run `ci/riscv-toolchain-install.sh`.

File: .github/workflows/benchmark.yml
Comment (jserv) [inline .github/workflows/benchmark.yml]: Specify `tests/benchmark_combiner.py` instead.

**Review by jserv — CHANGES_REQUESTED** [review]
Rename `tests/benchmark_combiner.py` to `tests/bench-aggregator.py`.

**Comment by ChinYikMing** [comment]
> Rename `tests/benchmark_combiner.py` to `tests/bench-aggregator.py`.

Done.

File: .github/workflows/benchmark.yml
Comment (jserv) [inline .github/workflows/benchmark.yml]: SDL is not required for building.

**Comment by jserv** [comment]
> Please add a secret called `RV32EMU_BENCH_TOKEN` to rv32emu repository, this token should be a PAT that having permission in scope of repo.

DONE.

> go to the setting of rv32emu repository > Action > General > make sure that "read and write" permission is granted in "Workflow permissions" section

DONE.

**Comment by jserv** [comment]
I defer to @qwe661234 for validating the scripts in `tests` directory.

File: .github/workflows/benchmark.yml
Comment (jserv) [inline .github/workflows/benchmark.yml]: Name it `bench-aggregator.py` without an underscore.

File: tests/bench-aggregator.py
Comment (jserv) [inline tests/bench-aggregator.py]: I suppose they are in the form of JSON, right? If so, use `json` as filename suffix.

Comment (ChinYikMing) [reply tests/bench-aggregator.py]: Yes

**Comment by qwe661234** [comment]
What information we would receive if the benchmark fails?

**Comment by ChinYikMing** [comment]
> What information we would receive if the benchmark fails?

Which part of failure in workflow that you referred to? 

Normally, if any failures occur during the workflow(i.e., failure of build rv32emu executable, segmentation fault, ...),  then the workflow will send an email to notify maintainer or committer. Of course, no new benchmark data will be updated in the web visualization of benchmark.

Originally, I want to comment the custom message in the thread of PR when unexpected condition happens, but it seems like this feature has not been implemented yet, see this [issue](https://github.com/benchmark-action/github-action-benchmark/issues/172)

**Comment by qwe661234** [comment]
> > What information we would receive if the benchmark fails?
> 
> Which part of failure in workflow that you referred to?
> 
> Normally, if any failures occur during the workflow(i.e., failure of build rv32emu executable, segmentation fault, ...), then the workflow will send an email to notify maintainer or committer. Of course, no new benchmark data will be updated in the web visualization of benchmark.
> 
> Originally, I want to comment the custom message in the thread of PR when unexpected condition happens, but it seems like this feature has not been implemented yet, see this [issue](https://github.com/benchmark-action/github-action-benchmark/issues/172)

Some errors only occur when running coreMark (e.g. JIT mode with clang compiler), so other workflows will pass and won't send emails to notify the maintainer or committer. This is the reason why I'm asking you this question, but it is not a critical problem.

**Review by qwe661234 — APPROVED** [review]

**Comment by jserv** [comment]
Thank @ChinYikMing for contributing!

**Comment by jserv** [comment]
> After merging this PR, please run the workflow called `Benchmark` manually via "Run workflow" button to generate the base benchmark. Note that it should be used in a reasonable timing in order not to mess up the benchmark visualization afterwards.
> Last, the benchmark CI should work!

@ChinYikMing, Please check the [CI results](https://github.com/sysprog21/rv32emu/actions/runs/6146294738) and add the hyperlink to benchmark visualization as a pull request for `README.md`.

**Pull request merged**
